crates/pipeline_manager/src/db/test.rs

```diff
@@ -0,0 +1,1140 @@
+    ),
+    AddPipelineToConfig(ConfigId, PipelineId),
+    RemovePipelineFromConfig(ConfigId),
+    UpdateConfig(
+        ConfigId,
+        Option<ProjectId>,
+        String,
+        String,
+        Option<String>,
+        Option<Vec<AttachedConnector>>,
+    ),
+    DeleteConfig(ConfigId),
+    NewPipeline(ConfigId, Version),
+    PipelineSetPort(PipelineId, u16),
+    SetPipelineKilled(PipelineId),
+    DeletePipeline(PipelineId),
+    GetPipeline(PipelineId),
+    ListPipelines,
+    NewConnector(String, String, ConnectorType, String),
+    ListConnectors,
+    GetConnector(ConnectorId),
+    UpdateConnector(ConnectorId, String, String, Option<String>),
+    DeleteConnector(ConnectorId),
+}
+
+fn check_responses<T: Debug + PartialEq>(step: usize, model: AnyResult<T>, impl_: AnyResult<T>) {
+    match (model, impl_) {
+        (Ok(mr), Ok(ir)) => assert_eq!(mr, ir),
+        (Err(me), Ok(ir)) => {
+            panic!("Step({step}): model returned error: {me:?}, but impl returned result: {ir:?}");
+        }
+        (Ok(mr), Err(ie)) => {
+            panic!("Step({step}): model returned result: {mr:?}, but impl returned error: {ie:?}");
+        }
+        (Err(me), Err(ie)) => {
+            assert_eq!(
+                me.to_string(),
+                ie.to_string(),
+                "Step({step}): Error return mismatch"
+            );
+        }
+    }
+}
+
+// Compare everything except the `created` field which gets set inside the DB.
+fn compare_pipeline(step: usize, model: AnyResult<PipelineDescr>, impl_: AnyResult<PipelineDescr>) {
+    match (model, impl_) {
+        (Ok(mut mr), Ok(mut ir)) => {
+            mr.created = DateTime::default();
+            ir.created = DateTime::default();
+            assert_eq!(mr, ir);
+        }
+        (Err(me), Ok(ir)) => {
+            panic!("Step({step}): model returned error: {me:?}, but impl returned result: {ir:?}");
+        }
+        (Ok(mr), Err(ie)) => {
+            panic!("Step({step}): model returned result: {mr:?}, but impl returned error: {ie:?}");
+        }
+        (Err(me), Err(ie)) => {
+            assert_eq!(
+                me.to_string(),
+                ie.to_string(),
+                "Step({step}): Error return mismatch"
+            );
+        }
+    }
+}
+
+/// Compare everything except tne `created` field which gets set inside the DB.
+fn compare_pipelines(
+    mut model_response: Vec<PipelineDescr>,
+    mut impl_response: Vec<PipelineDescr>,
+) {
+    assert_eq!(
+        model_response
+            .iter_mut()
+            .map(|p| p.created = DateTime::default())
+            .collect::<Vec<_>>(),
+        impl_response
+            .iter_mut()
+            .map(|p| p.created = DateTime::default())
+            .collect::<Vec<_>>()
+    );
+}
+
+/// Compare the database storage implementation with our model using in-memory
+/// data-structures. Ideally, the two behave the same.
+#[test]
+fn db_impl_behaves_like_model() {
```

Comment:
Why are we maintaining a model in addition to the DB?

Reply:
Now we have good test coverage ;) -- ideally the model could be less code, but it did find a number of bugs in the impl while writing it so that was good

Reply:
I mean, if we can write a test against the model, we should just as well write that test against the DB no?

Reply:
Not sure I follow; every sequence of actions is evaluated against both implementations model  PostgresDB?. The test makes sure they return the same things.

Reply:
Ah, got it. Sounds good. We'll need to plan for transactions/aborts/rollbacks at some point, but this is a good way to catch bugs as you say.